api/src/services/users.ts

```diff
@@ -451,6 +458,118 @@ export class UsersService extends ItemsService {
 		await service.updateOne(user.id, { password, status: 'active' });
 	}
 
+	async registerUser(input: RegisterUserInput) {
+		const STALL_TIME = 750;
+		const timeStart = performance.now();
+		const serviceOptions: AbstractServiceOptions = { accountability: this.accountability, schema: this.schema };
+		const settingsService = new SettingsService(serviceOptions);
+
+		const settings = await settingsService.readSingleton({
+			fields: [
+				'is_public_registration_enabled',
+				'is_public_registration_email_validation_enabled',
+				'public_registration_role',
+				'public_registration_email_filter',
+			],
+		});
+
+		if (settings?.['is_public_registration_enabled'] == false) {
+			await stall(STALL_TIME, timeStart);
+			throw new ForbiddenError();
+		}
+
+		const publicRegistrationRole = settings?.['public_registration_role'];
+
+		if (!publicRegistrationRole) {
+			await stall(STALL_TIME, timeStart);
+			throw new ContainsNullValuesError({ collection: 'directus_settings', field: 'public_registration_role' });
+		}
+
+		const hasEmailValidation = settings?.['is_public_registration_email_validation_enabled'];
+
+		const partialUser: Partial<User> = {
+			// Required fields
+			email: input.email,
+			password: input.password,
+			role: publicRegistrationRole,
+			status: hasEmailValidation ? 'draft' : 'active', // TODO: Do we want to have a dedicated "unverified" status?
+
+			// Optional fields
+			first_name: input.first_name ?? null,
+			last_name: input.last_name ?? null,
+		};
+
+		const emailFilter = settings?.['public_registration_email_filter'];
+
+		if (hasEmailValidation && emailFilter && validatePayload(emailFilter, { email: input.email }).length !== 0) {
+			await stall(STALL_TIME, timeStart);
+			throw new FailedValidationError({ field: 'email', type: 'email' });
+		}
+
+		try {
+			await this.createOne(partialUser);
+		} catch (error: unknown) {
+			// To avoid giving attackers infos about registered emails we dont fail for violated unique constraints
+			if (isDirectusError(error) && error.code === 'RECORD_NOT_UNIQUE') {
+				await stall(STALL_TIME, timeStart);
+				return;
+			}
+
+			await stall(STALL_TIME, timeStart);
+			throw error;
+		}
+
+		if (hasEmailValidation) {
+			const mailService = new MailService(serviceOptions);
+			const payload = { email: input.email, scope: 'pending-registration' };
+			const token = jwt.sign(payload, env['SECRET'] as string, { expiresIn: '7d', issuer: 'directus' });
```

Comment:
Mhh initially I agreed because I thought we can reuse it for more tokens but now I looked up our other TTLs and theres no further overlap, just the invite and registration. Furthermore it sounds like this could be a nice configuration env variable even! For example we also have `SESSION_COOKIE_TTL`, `ACCESS_TOKEN_TTL` so why not add something like `EMAIL_VERIFICATION_TOKEN_TTL`, I can already see other people wanting to override this to shorter time frames. I have not thought about this initially, thanks! :heart:  :+1: